sdk/core/azure-core/inc/azure/core/internal/client_options.hpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Base type for all client option types, exposes various common client options like Retry
+ * and Transport.
+ */
+
+#pragma once
+
+#include "azure/core/http/http.hpp"
+#include "azure/core/http/policy.hpp"
+
+#include <memory>
+#include <vector>
+
+namespace Azure { namespace Core { namespace Internal {
+
+  /**
+   * @brief Represents a position of the policy in the pipeline.
+   *
+   */
+  enum class HttpPipelinePosition
+  {
+    /**
+     * @brief The policy would be invoked once per pipeline invocation (service call).
+     *
+     */
+    PerCall,
+    /**
+     * @brief The policy would be invoked every time request is retried.
+     *
+     */
+    PerRetry
+  };
+
+  /**
+   * @brief  Base type for all client option types, exposes various common client options like Retry
+   * and Transport.
+   *
+   */
+  class ClientOptions {
+  private:
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perOperationPolicies;
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perRetryPolicies;
+
+  public:
+    /**
```

Comment:
why do we need all these constructors, and member functions? Why not just

```
struct ClientOptions {
    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perOperationPolicies;
    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perRetryPolicies;
    Azure::Core::Http::TransportPolicyOptions Transport;
    ...
};
```

Reply:
The per operation and per retry are private, once they are added we don't want them public

Reply:
I don't quite understand why we need to do that. To me, the policies are just the same as other options like `RetryOptions` and `TelemetryPolicyOptions`. If we want to prevent customers from modifying policies, why don't we need to prevent them from modifying other options?

Reply:
>If we want to prevent customers from modifying policies

What we want is to encapsulate the feature of `policies extensibility`.  The feature allow customers to add custom policies to the pipeline by calling `AddPolicy()` from any `clientOptions`. 

The problem with making the vector<policies> public is that the implementation details of the feature becomes the exposed and never to be changed. We might want to change the vector<> for something else (the implementation) and the way to do it is by encapsulating.  The customers just calls AddPolicy and set if they want it per call or per retry only.

Now, we can think about the policies as 3 types:
- Default: The ones mandated by Azure Core (like retry, telemetry, transport, etc). This policies ensure all SDK clients offers the same functionalities to customers. This policies are added by Azure Core and set it up through PolicyOptions as input. The policy options are in the ClientOptions base class.

- Service's policies:  These are specific to services and works together with the default ones. Same as the default policies, the customers never interact with the policy but with the policyOptions only. The options for this policies are in the clientOptions' derived class (like blobClientOptions).  Azure Core will never know about these options. The Service will use the options from the derived class to create the policies and will pass them to Azure Core to use them next to the default.

- Customer policies. Policies created by a customer and attached through the AddPolicy() API exposed by the clientOptions base class. The services will never know about these policies.

The new constructor for the Http pipeline takes the 3 types of policies to build the Pipeline for the service client.

Reply:
okay, although I don't like this way. But it doesn't hurt me anyway.